File: tests/unit/utils/formatting.test.ts
import { describe, it, expect } from "vitest";
import {
  escapeMarkdownV2,
  escapeHtml,
  bold,
  italic,
  code,
  link,
  userMention,
  spoiler,
  strikethrough,
  underline,
  blockquote,
  expandableBlockquote,
} from "../../../src/utils/formatting.js";

describe("Formatting & Escaping Utilities", () => {
  it("escapeMarkdownV2 escapes all Telegram MarkdownV2 special characters", () => {
    const raw = "Hello *world*! [Link](url) _test_ ~strike~ `code` >quote #tag + - = | { } . !";
    const escaped = escapeMarkdownV2(raw);
    expect(escaped).toContain("\\*");
    expect(escaped).toContain("\\[");
    expect(escaped).toContain("\\_");
    expect(escaped).toContain("\\!");
    expect(escaped).toContain("\\.");
  });

  it("escapeHtml escapes HTML dangerous entities", () => {
    const raw = '<script>alert("test & win")</script>';
    const escaped = escapeHtml(raw);
    expect(escaped).toBe("&lt;script&gt;alert(&quot;test &amp; win&quot;)&lt;/script&gt;");
  });

  it("formats bold text in HTML and MarkdownV2", () => {
    expect(bold("Alert!")).toBe("<b>Alert!</b>");
    expect(bold("Alert!", "MarkdownV2")).toBe("*Alert\\!*");
  });

  it("formats italic text in HTML and MarkdownV2", () => {
    expect(italic("Notice")).toBe("<i>Notice</i>");
    expect(italic("Notice", "MarkdownV2")).toBe("_Notice_");
  });

  it("formats inline and multi-line code blocks", () => {
    expect(code("const x = 1;")).toBe("<code>const x = 1;</code>");
    expect(code("const x = 1;", "typescript")).toBe(
      '<pre><code class="language-typescript">const x = 1;</code></pre>',
    );
    expect(code("const x = 1;", "typescript", "MarkdownV2")).toContain("```typescript\n");
  });

  it("formats links and user mentions", () => {
    expect(link("Docs", "https://t.me")).toBe('<a href="https://t.me">Docs</a>');
    expect(userMention("Alice", 123456)).toBe('<a href="tg://user?id=123456">Alice</a>');
  });

  it("formats spoilers, strikethrough, underline, and blockquotes", () => {
    expect(spoiler("Secret")).toBe("<tg-spoiler>Secret</tg-spoiler>");
    expect(strikethrough("Old")).toBe("<s>Old</s>");
    expect(underline("Underlined")).toBe("<u>Underlined</u>");
    expect(blockquote("Quote text")).toBe("<blockquote>Quote text</blockquote>");
    expect(expandableBlockquote("Long text")).toBe("<blockquote expandable>Long text</blockquote>");
  });
});

File: packages/go/pkg/utils/formatting_test.go
package utils_test

import (
	"strings"
	"testing"

	"github.com/Nam088/telebot-go/pkg/utils"
)

func TestEscapeMarkdownV2(t *testing.T) {
	// Every MarkdownV2-reserved character, escaped with a leading backslash.
	reserved := `_*[]()~` + "`" + `>#+-=|{}.!\/`
	var escapedReserved strings.Builder
	for _, r := range reserved {
		escapedReserved.WriteByte('\\')
		escapedReserved.WriteRune(r)
	}

	tests := []struct {
		name string
		in   string
		want string
	}{
		{"empty", "", ""},
		{"plain text", "hello world", "hello world"},
		{"price and brackets", "Price: $10.00! [Sale]", `Price: $10\.00\! \[Sale\]`},
		{"all reserved chars", reserved, escapedReserved.String()},
		{"unicode preserved", "Привет, мир! 🚀", `Привет, мир\! 🚀`},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := utils.EscapeMarkdownV2(tt.in); got != tt.want {
				t.Errorf("EscapeMarkdownV2(%q) = %q, want %q", tt.in, got, tt.want)
			}
		})
	}
}

func TestEscapeHTML(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{"empty", "", ""},
		{"plain text", "hello world", "hello world"},
		{"script tag", "<script>alert('xss') & win</script>", "&lt;script&gt;alert(&#39;xss&#39;) &amp; win&lt;/script&gt;"},
		{"double quotes", `She said "hi"`, "She said &quot;hi&quot;"},
		{"unicode preserved", "Ärger & Spaß", "Ärger &amp; Spaß"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := utils.EscapeHTML(tt.in); got != tt.want {
				t.Errorf("EscapeHTML(%q) = %q, want %q", tt.in, got, tt.want)
			}
		})
	}
}

func TestFormattingHelpersHTML(t *testing.T) {
	tests := []struct {
		name string
		got  string
		want string
	}{
		{"bold", utils.Bold("a<b", utils.ModeHTML), "<b>a&lt;b</b>"},
		{"italic", utils.Italic("hi", utils.ModeHTML), "<i>hi</i>"},
		{"code inline", utils.Code("x=1", "", utils.ModeHTML), "<code>x=1</code>"},
		{"code language", utils.Code("let a = 1", "typescript", utils.ModeHTML),
			`<pre><code class="language-typescript">let a = 1</code></pre>`},
		{"code language escaped", utils.Code("body", "c++", utils.ModeHTML),
			`<pre><code class="language-c++">body</code></pre>`},
		{"link", utils.Link("Site", "https://example.com/?a&b", utils.ModeHTML),
			`<a href="https://example.com/?a&amp;b">Site</a>`},
		{"mention", utils.UserMention("Alice", 42, utils.ModeHTML),
			`<a href="tg://user?id=42">Alice</a>`},
		{"spoiler", utils.Spoiler("secret", utils.ModeHTML), "<tg-spoiler>secret</tg-spoiler>"},
		{"strike", utils.Strike("old", utils.ModeHTML), "<s>old</s>"},
		{"underline", utils.Underline("important", utils.ModeHTML), "<u>important</u>"},
		{"blockquote", utils.Blockquote("quote", utils.ModeHTML), "<blockquote>quote</blockquote>"},
		{"expandable blockquote", utils.ExpandableBlockquote("quote", utils.ModeHTML),
			"<blockquote expandable>quote</blockquote>"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.got != tt.want {
				t.Errorf("got %q, want %q", tt.got, tt.want)
			}
		})
	}
}

func TestFormattingHelpersMarkdownV2(t *testing.T) {
	tests := []struct {
		name string
		got  string
		want string
	}{
		{"bold", utils.Bold("a.b", utils.ModeMarkdownV2), `*a\.b*`},
		{"italic", utils.Italic("hi!", utils.ModeMarkdownV2), `_hi\!_`},
		// Inside code spans only backtick and backslash are escaped.
		{"code inline partial escape", utils.Code("a*b`c\\d", "", utils.ModeMarkdownV2),
			"`a*b\\`c\\\\d`"},
		{"code language", utils.Code("x := 1`", "go", utils.ModeMarkdownV2), "```go\nx := 1\\`\n```"},
		{"code language escaped", utils.Code("body", "c++", utils.ModeMarkdownV2), "```c\\+\\+\nbody\n```"},
		{"link", utils.Link("My Site", "https://example.com/a_b", utils.ModeMarkdownV2),
			`[My Site](https:\/\/example\.com\/a\_b)`},
		{"mention", utils.UserMention("Alice", 42, utils.ModeMarkdownV2),
			`[Alice](tg:\/\/user?id\=42)`},
		{"spoiler", utils.Spoiler("secret.", utils.ModeMarkdownV2), `||secret\.||`},
		{"strike", utils.Strike("old", utils.ModeMarkdownV2), `~old~`},
		{"underline", utils.Underline("important", utils.ModeMarkdownV2), `__important__`},
		{"blockquote", utils.Blockquote("quote", utils.ModeMarkdownV2), `>quote`},
		{"expandable blockquote", utils.ExpandableBlockquote("quote", utils.ModeMarkdownV2), `**>quote||`},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.got != tt.want {
				t.Errorf("got %q, want %q", tt.got, tt.want)
			}
		})
	}
}

func TestUnknownModeFallsBackToMarkdownV2(t *testing.T) {
	if got, want := utils.Bold("hi", "invalid"), `*hi*`; got != want {
		t.Errorf("Bold with unknown mode = %q, want %q", got, want)
	}
	if got, want := utils.Spoiler("hi", ""), "||hi||"; got != want {
		t.Errorf("Spoiler with empty mode = %q, want %q", got, want)
	}
}
